alacritty/src/display/mod.rs

```diff
@@ -980,74 +1020,203 @@ impl Display {
+                };
+
+                let shortener_position = if num_chars < max_width {
+                    // Allocate space for the char we're about to write.
+                    visible_text.push(' ');
+                    num_chars
+                } else if num_chars > 1
+                    && visible_text.get(num_chars - 2).and_then(|ch| ch.width()) == Some(2)
+                {
+                    // Remove the trailing space.
+                    let _ = visible_text.pop();
+                    num_chars - 2
+                } else {
+                    num_chars - 1
+                };
+
+                visible_text[shortener_position] = shortener;
+
+                break;
             }
+
+            visible_text.push(ch);
+
+            // Add spacers for wide chars.
+            if char_width == 2 {
+                visible_text.push(' ');
+            }
+
+            num_chars += char_width;
         }
 
-        // Add cursor to show whitespace.
-        formatted_regex.push('_');
+        visible_text
+    }
 
-        // Truncate beginning of the search regex if it exceeds the viewport width.
-        let num_cols = size_info.columns();
-        let label_len = search_label.chars().count();
-        let regex_len = formatted_regex.chars().count();
-        let truncate_len = cmp::min((regex_len + label_len).saturating_sub(num_cols), regex_len);
-        let index = formatted_regex.char_indices().nth(truncate_len).map(|(i, _c)| i).unwrap_or(0);
-        let truncated_regex = &formatted_regex[index..];
+    /// Draw preview for the currently highlighted `Hyperlink`.
+    #[inline(never)]
+    fn draw_hyperlink_preview(
+        &mut self,
+        config: &UiConfig,
+        vi_cursor_point: Option<Point>,
+        display_offset: usize,
+    ) {
+        let num_cols = self.size_info.columns();
+        let uris: Vec<_> = self
+            .highlighted_hint
+            .iter()
+            .chain(&self.vi_highlighted_hint)
+            .filter_map(|hint| hint.hyperlink().map(|hyperlink| hyperlink.uri()))
+            .map(|uri| {
+                Self::format_visible_text(uri.chars(), num_cols, Some(URI_SHORTENER)).into_iter()
+            })
+            .collect();
+
+        if uris.is_empty() {
+            return;
+        }
 
-        // Add search label to the beginning of the search regex.
-        let mut bar_text = format!("{}{}", search_label, truncated_regex);
+        // The maximum amount of protected lines including the ones we'll show preview on.
+        let max_protected_lines = uris.len() * 2;
+
+        // Lines we shouldn't shouldn't show preview on, because it'll obscure the highlighted
+        // hint.
+        let mut protected_lines = Vec::with_capacity(max_protected_lines);
+        if self.size_info.screen_lines() >= max_protected_lines {
+            // Prefer to show preview even when it'll likely obscure the highlighted hint, when
+            // there's no place left for it.
```

Comment:
I've tested this and when I have a two-line terminal with search and hover the URL, I don't get any preview.